src/widgets/modals/ByProgramModal.js

```diff
@@ -0,0 +1,406 @@
+      break;
+    case 'SELECT_NAME':
+    case 'SELECT_PERIOD':
+    case 'SELECT_SUPPLIER':
+      newState = { modalIsOpen: false };
+      break;
+    default:
+      break;
+  }
+
+  return newState;
+};
+
+export class ByProgramModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      // Values used for deriving current UI state
+      currentStep: 0,
+      steps: null,
+      modalIsOpen: false,
+      isProgramBased: true,
+      // New requisition or stocktake values
+      program: {},
+      supplier: {},
+      orderType: {},
+      period: {},
+      name: null,
+      // Data stored for performance
+      storeTag: null,
+      programs: null,
+      suppliers: null,
+      orderTypes: null,
+      periods: null,
+    };
+  }
+
+  // Store unchanging select modal options and set the default
+  // steps array.
+  componentDidMount = () => {
+    const { settings, database } = this.props;
+    const programs = getAllPrograms(settings, database);
+    const suppliers = database.objects('Name').filtered('isSupplier = $0', true);
+    this.setCurrentSteps();
+    this.setState({ programs, suppliers });
+  };
+
+  getProgramValues = () => {
+    const { program, supplier, orderType, period, name } = this.state;
+    return { program, supplier, orderType, period, name };
+  };
+
+  getStepNumbers = () => {
+    const { isProgramBased } = this.state;
+    return {
+      program: 0,
+      supplier: isProgramBased ? 1 : 0,
+      orderType: 2,
+      period: 3,
+      name: isProgramBased ? 1 : 0,
+    };
+  };
+
+  // Open a modal - in case it is an earlier step in the process,
+  // set the current step to the step number corresponding with
+  // the onPress event.
+  openModal = ({ stepNumber }) => {
+    this.setState({ modalIsOpen: true, currentStep: stepNumber });
+  };
+
+  closeModal = () => {
+    this.setState({ modalIsOpen: false });
+  };
+
+  // Initial set up or update the current steps property to reflect the current state
+  setCurrentSteps = () => {
+    const sequentialSteps = this.getSequentialStepsProps();
+    const steps = this.getSteps().map(step => sequentialSteps[step]);
+    this.setState({ steps });
+  };
+
+  // Gets the available steps this modal can handle
+  getSteps = () => {
+    const { isProgramBased } = this.state;
+    const { type } = this.props;
+    let steps;
+    switch (type) {
+      case 'requisition':
+        if (isProgramBased) steps = ['program', 'supplier', 'orderType', 'period'];
+        else steps = ['supplier'];
+        break;
+      case 'stocktake':
+        if (isProgramBased) steps = ['program', 'name'];
+        else steps = ['name'];
+        break;
+      default:
+        break;
+    }
+    return steps;
+  };
+
+  getSequentialStepsProps = () => {
+    const { name } = this.state;
+    const stepNumbers = this.getStepNumbers();
+    const programValues = this.getProgramValues();
```

Comment:
maybe program values should be in state.programValues ? this way don't need this.getProgramValues method